[core] Remove `withTheme` from `@material-ui/core` (#26051)
mui-org/material-ui: docs/src/pages/guides/migration-v4/migration-v4.md

### withTheme

- The `withTheme` HOC utility has been removed from the `@material-ui/core/styles` package. You can use `@material-ui/styles/withTheme` instead. Make sure to add a `ThemeProvider` at the root of your application, as the `defaultTheme` is no longer available. If you are using this utility together with `@material-ui/core`, it's recommended you use the `ThemeProvider` component from `@material-ui/core/styles` instead.

  ```diff

>   - import { withTheme } from '@material-ui/core/styles';

Reviewer: It renders funny:

```suggestion
  -import { withTheme } from '@material-ui/core/styles';
```

<img width="163" alt="Screenshot 2021-04-29 at 20 23 38" src="https://user-images.githubusercontent.com/3165635/116599780-d149e900-a928-11eb-9aac-0c2875579b4d.png">

Author: Updated 👍

### withTheme

> - The `withTheme` utility has been removed from the `@material-ui/cire/styles` package. You can use the the `@material-ui/styles/withTheme` instead. Make sure to add a `ThemeProvider` at the root of your application, as the `defaultTheme` is no longer available. If you are using this utility together with `@material-ui/core`, it's recommend you use the `ThemeProvider` from `@material-ui/core/styles` instead.

Reviewer: ```suggestion
- The `withTheme` HOC utility has been removed from the `@material-ui/core/styles` package. You can use `@material-ui/styles/withTheme` instead. Make sure to add a `ThemeProvider` at the root of your application, as the `defaultTheme` is no longer available. If you are using this utility together with `@material-ui/core`, it's recommended you use the `ThemeProvider` component from `@material-ui/core/styles` instead.
```

  ```diff

>     import * as React from 'react';

Reviewer: noise

```suggestion
```

  - import { withTheme } from '@material-ui/core/styles';
  + import { withTheme } from '@material-ui/styles';

>   + import { createMuiTheme, ThemeProvider } from '@material-ui/core/styles';

Reviewer: ```suggestion
  + import { createTheme, ThemeProvider } from '@material-ui/core/styles';
```

>   +  const theme = createMuiTheme();

Reviewer: ```suggestion
  + const theme = createTheme();
```

    const MyComponent = withTheme(({ theme }) => <div>{props.theme.direction}</div>);

>     function App(props) {

Reviewer: ```suggestion
   function App(props) {
```

  -  return <MyComponent />;
  +  return <ThemeProvider theme={theme}><MyComponent {...props} /></ThemeProvider>;

>     }

Reviewer: ```suggestion
   }
```
